File: testerum-backend/testerum-runner/testerum-runner-cmdline/src/main/assembly/report_templates/common/testerum-model/model/step/def/composed-step-def.ts
import {StepDef} from "./step-def";
import {Path} from "../../path";
import {StepPhaseEnum} from "./step-phase-enum";
import {StepPattern} from "./pattern/step-pattern";
import {StepCall} from "../call/step-call";
import {MarshallingUtils} from "../../../json-marshalling/marshalling-utils";

export class ComposedStepDef implements StepDef {

    constructor(public readonly id: string,
                public readonly path: Path,
                public readonly oldPath: Path | null,
                public readonly phase: StepPhaseEnum,
                public readonly stepPattern: StepPattern,
                public readonly description: string | null,
                public readonly tags: Array<string>,
                public readonly stepCalls: Array<StepCall>) { }

    static parse(input: Object): ComposedStepDef {
        const id = input["id"];
        const path = Path.parse(input["path"]);
        const oldPath = Path.parse(input["oldPath"]);
        const phase = MarshallingUtils.parseEnum(input["phase"], StepPhaseEnum);
        const stepPattern = StepPattern.parse(input["stepPattern"]);
        const description = input["description"];
        const tags = MarshallingUtils.parseListOfStrings(input["tags"]);
        const stepCalls = MarshallingUtils.parseList(input["stepCalls"], StepCall);

        return new ComposedStepDef(id, path, oldPath, phase, stepPattern, description, tags, stepCalls);
    }
}

File: old/model/src/main/kotlin/com/testerum/model/step/ComposedStepDef.kt
package com.testerum.model.step

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import com.testerum.model.enums.StepPhaseEnum
import com.testerum.model.infrastructure.path.Path
import com.testerum.model.text.StepPattern
import com.testerum.model.util.StepHashUtil
import com.testerum.model.warning.Warning

data class ComposedStepDef @JsonCreator constructor(@JsonProperty("path")        override val path: Path,
                                                    @JsonProperty("oldPath")     val oldPath: Path? = path,
                                                    @JsonProperty("phase")       override val phase: StepPhaseEnum,
                                                    @JsonProperty("stepPattern") override val stepPattern: StepPattern,
                                                    @JsonProperty("description") override val description: String? = null,
                                                    @JsonProperty("tags")        val tags: List<String> = emptyList(),
                                                    @JsonProperty("stepCalls")   val stepCalls: List<StepCall>,
                                                    @JsonProperty("warnings")    override val warnings: List<Warning> = emptyList(),
                                                    @JsonProperty("isUsed")      val isUsed: Boolean = true): StepDef {

    companion object {
        val COMPOSED_STEP_FILE_EXTENSION = "step"
    }

    override val id: String
        get() = StepHashUtil.calculateStepHash(phase, stepPattern)

    private val _descendantsHaveWarnings: Boolean = stepCalls.any { it.warnings.isNotEmpty() || it.descendantsHaveWarnings }

    @get:JsonProperty("descendantsHaveWarnings")
    override val descendantsHaveWarnings: Boolean
        get() = _descendantsHaveWarnings

    @JsonIgnore
    fun getNewPath(): Path {
        return path.copy(
                fileName = getText(),
                fileExtension = COMPOSED_STEP_FILE_EXTENSION
        )
    }

    override fun toString() = buildString {
        append(phase)
        append(" ")
        append(stepPattern.toDebugString(varPrefix = "<<", varSuffix = ">>"))
        append(" (").append(stepCalls.size).append(" calls)")
    }
}
